### ISSUE TITLE
Federation + Dataloaders/batching

### ISSUE THREAD

**Issue by BrendanBall** [apollographql/apollo-server#2887]
Feature request

Federation looks really nice, but according to the docs there doesn't seem to be a way to incorporate dataloaders into it and there is no mention of federation implementing any kind of batching and how it would do that. Currently we have special batch queries used in schema stitching to efficiently load a long list of items, so how would this be done in federation?

**Comment by ksaldana1** [comment]
I have thought about this a bit and have a few thoughts and temporary workarounds. I agree that it would be nice for the Gateway to incorporate some sort of batching mechanism—but it doesn't seem like that's the case today. So the question is: if we can't leverage batching at the Gateway level over the wire, how can we still use Dataloaders at the individual federated service level  (to minimize duplicated database work). 

The problem you have to solve is: given that a single "request" can now be multiple calls over the wire to individual services, how can I group together those multiple calls. Before you could just new up your Dataloaders inside of your context initializer, which is how you maintained request scoped context. This is no longer the case. You need some other unique identifier to recognize the scope of each request. For us, the user access token on headers is how we are scoping these requests.

Because multiple calls for one GraphQL query can hit individual services, your Dataloders for references can no longer live in context initialization. It has to live outside that in memory—tldr you need a cache of caches. Hacky bullshit code incoming:

```typescript
class UserReferenceLoader {
  private cache: Record<string, DataLoader<string, avero.user.User>>;

  constructor() {
    this.cache = {};
    this.clearInterval();
  }

  private clearInterval() {
    setInterval(() => {
      this.cache = {};
    }, 20000);
  }

  create = (token: string) => {
    if (this.cache[token]) {
      return this.cache[token];
    }
    this.cache[token] = new DataLoader(async (usernames: string[]) => {
      const { users } = await Users.getUsers(
        { usernames },
        toGrpcMetadata({ token })
      ).toPromise();
      return usernames.map(username => users.find(user => user.username === username));
    });
    return this.cache[token];
  };
}
```
and now in the resolvers: 
```typescript
__resolveReference: (user, { token, UserReferenceLoader }) => {
      return UserReferenceLoader.create(token).load(user.username);
    }
```
This doesn't feel great, and I'm sure there are some gaping holes in design here, but the basic idea works—this batches together User resolutions for a single GraphQL request that is backed by multiple calls to the User federated service.

Would definitely be nice to have something at the Gateway layer instead. This is definitely bleeding the internals of the service mesh into my individual services.

**Comment by j** [comment]
I was wondering this same thing.

If a `Post` service has a field to get the `author` from the `User`, and you're returning a list of posts, does the current implementation do many requests to the `User` service or combine them to one?  If it's combined to one, how can you do one database call on the list of user ids?

**Comment by fernandobandeira** [comment]
Guys it's simpler than that, federation sends a single request to the service for each query, so in the example above by @j you start by creating a function that returns a dataloader:

```javascript
const getTransactionLoader = () => new DataLoader(
  async ids => {
    const transactions = await Transaction.find({ _id: { $in: ids } });
    const transactionsById = keyBy(transactions, "id");
    return ids.map(id => transactionsById[id]);
  }
)
```

Add the loader instance to the context:

```javascript
context: ({ req }) => ({
    transactionLoader: getTransactionLoader(),
)}
```

Use the dataloader

```javascript
__resolveReference: ({ id }, { transactionLoader }) => transactionLoader.load(id),
```

this will group the ids for this context, since a context can't have multiple users there's no need to consider the token when grouping the ids.

**Comment by BrendanBall** [comment]
@fernandobandeira can you explain how this works that it's able to send a single request for each query? Currently the query planner is broken so I can't see this in action and I haven't found any docs that explain this. Based on the federation spec I can kind of see how it would work.

so if I had a query
```
{
  posts {
   title
   author {
     name
   }
  }
}
```
where `posts` is a service and `author` returns an `account` from the `accounts` service. Would the query plan be something like this:
```
{
  Fetch(service: "posts") {
    posts {
      title
      author {
        __typename
        id
      }
    }
  }
  Fetch(service: "accounts") {
    _entities(representations: $ListOfAccountsWithIds) {
     ... on Account {
      id
      name
    }
   }
  }
}
```
such that request 1 to posts service returns:
```
{
  posts: [
   {
     title: "Post 1",
     author: {
        __typename: "Account",
        id: 1
     }
   },
   {
     title: "Post 2",
     author: {
        __typename: "Account",
        id: 2
     }
   }
  ]
}
```
and request 2 to accounts service returns:
```
{
  _entities: [
    {
        __typename: "Account",
        id: 1,
        name: "Bob"
     },
    {
        __typename: "Account",
        id: 2,
        name: "Alice"
     }
  ]
}
```
Is this correct?

**Comment by ksaldana1** [comment]
@fernandobandeira That is true if all of your resolutions are coming from a singular service—that does not allow you to group multiple reference resolutions that span across multiple federated services. But even looking at network traffic, that invariant does not seem to always hold true.

I have added a `console.log` to the UserLoader as in your example above. When sending a request like this
```
{
  self {
    size1: kitchens {
      getCounts(input: { page: { size: 1, number: 1 } }) {
        counts {
          createdBy {
            firstName
            lastName
          }
        }
      }
    }
    size3: kitchens {
      getCounts(input: { page: { size: 3, number: 2 } }) {
        counts {
          createdBy {
            firstName
            lastName
          }
        }
      }
    }
  }
}
```

<img width="478" alt="Screen Shot 2019-06-25 at 8 51 27 AM" src="https://user-images.githubusercontent.com/12038627/60108770-72145f80-9726-11e9-8f8d-35f48891ecc5.png">

Two network hops and two database requests despite the fact that these resolutions are happening from the same service. If I expand to multiple services, the behavior worsens.

```
query {
  self {
    kitchens {
      id
      name
      getCounts(input: { page: { number: 1, size: 5 } }) {
        counts {
          createdBy {
            username
            firstName
            lastName
          }
        }
      }
    }
  	details {
      organizations(ids: ["7047"]){
        id
        logbook {
          id
          name
          page(day: "2019-05-29") {
            sections {
              fields {
                entries {
                  ... on TextEntry {
                    updatedBy {
                      firstName
                      lastName
                    }
                  }
                  ...on BooleanEntry {
                    updatedBy {
                      firstName
                      lastName
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

<img width="483" alt="Screen Shot 2019-06-25 at 8 52 42 AM" src="https://user-images.githubusercontent.com/12038627/60108880-9c661d00-9726-11e9-9a85-800098f47f7f.png">

3 network hops to resolve users, multiple duplicates. 

Using my token (cache of caches), I somewhat alleviate this.

<img width="480" alt="Screen Shot 2019-06-25 at 8 53 26 AM" src="https://user-images.githubusercontent.com/12038627/60108968-bdc70900-9726-11e9-9078-ab30d5035342.png">

There is still 2 requests happening due to the timing and async nature of all this, but it's definitely doing something. I think the OPs request is still valid and there's some questions on best practices around batching to be done.

**Comment by fernandobandeira** [comment]
@BrendanBall 

Yes, that's correct, check if you have tracing enabled on any of your apollo servers, disabling them should make the query plan appear on playground, also make sure that the node_env isn't set to production.

Here's how the relationship query plan looks like:

```javascript
Flatten(path: "posts.@") {
      Fetch(service: "accounts") {
        {
          ... on Account {
            __typename
            id
          }
        } =>
        {
          ... on Account {
            name
          }
        }
      },
    },
```

@ksaldana1 Your point is indeed true, but if you have your users services behind a load balancer on prod with multiple replicas each request from those other services might go to different services underneath so the trouble to optimize this won't be that useful, I think that you'd be better using `apollo-server-plugin-response-cache`, this way you'd cache the response and would hit the database two times like your approach is doing even if the requests are split between the replicas.

**Comment by BrendanBall** [comment]
@fernandobandeira awesome, please see [issue 3](https://github.com/apollographql/federation-demo/issues/3)

**Comment by j** [comment]
Awesome... now only if GraphQL + Typescript (or plain JS) wasn't schema-first and had federation support.. I'd be in heaven.

**Comment by martijnwalraven** [comment]
We should do a better job explaining this in the docs, but gateway batches requests to underlying services by default. The query planner will generate a single fetch for all entities at a certain level.

So for @BrendanBall's example in https://github.com/apollographql/apollo-server/issues/2887#issuecomment-505479000:
```
{
  posts {
   title
   author {
     name
   }
  }
}
```
We would indeed make two request: one to the blogging service to get the list of posts and the author IDs, and one to the accounts service to get the names of all the authors by ID.

That doesn't solve the problem of batching request to underlying data sources (like databases or REST endpoints) though. So you'll still want to use data loader (in `__resolveReference`) to make sure you can efficiently load a lists of posts or users in each individual service.
